src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -752,7 +753,11 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                         break;
                 }
 
-                Debug.Assert((object)this._type != null);
+                if ((object)this._type == null)
+                {
+                    SetType(_nodeBinder.CreateErrorType("var"));
```

Comment:
>SetType(_nodeBinder.CreateErrorType("var")); [](start = 20, length = 44)

> There isn't a mechanism to "enforce" that we only get here for broken code, because we don't have the necessary compilation diagnostics in hand. Ideas for doing that are welcome.

I believe we should never get here for an Out Var or a pattern variable. It should be fairly easy to detect such cases by examining the syntax. Please add the detection and a throw for them.